samples/nrf9160/modem_shell/src/sock/sock.c

```diff
@@ -0,0 +1,1320 @@
+			shell_print(
+				shell_global,
+				"socket send failed, err %d",
+				errno);
+		}
+		return -1;
+	}
+	return bytes;
+}
+
+static void data_send_work_handler(struct k_work *item)
+{
+	struct data_transfer_info *data_send_info_ptr =
+		CONTAINER_OF(item, struct data_transfer_info, work);
+	struct sock_info_t *socket_info = data_send_info_ptr->parent;
+
+	if (!socket_info->in_use) {
+		shell_print(
+			shell_global,
+			"Socket id=%d not in use. Fatal error and sending won't work.",
+			socket_info->id);
+		/* TODO: stop timer */
+		return;
+	}
+
+	sock_send(
+		socket_info,
+		socket_info->send_buffer,
+		socket_info->send_buffer_size,
+		true,
+		data_send_info_ptr->data_format_hex);
+}
+
+static void data_send_timer_handler(struct k_timer *dummy)
+{
+	struct data_transfer_info *data_send_info_ptr =
+		CONTAINER_OF(dummy, struct data_transfer_info, timer);
+	struct sock_info_t *socket_info = data_send_info_ptr->parent;
+
+	k_work_submit(&socket_info->send_info.work);
+}
+
+static void sock_send_random_data_length(struct sock_info_t *socket_info)
+{
+	int bytes;
+	int64_t time_intermediate;
+	int64_t ul_time_intermediate_ms;
+	double throughput;
+	int64_t ul_time_ms;
+
+	while (socket_info->send_bytes_left > 0) {
+		if (socket_info->send_bytes_left < socket_info->send_buffer_size) {
+			memset(socket_info->send_buffer, 0, socket_info->send_buffer_size);
+			memset(socket_info->send_buffer, 'l', socket_info->send_bytes_left);
+		}
+		bytes = sock_send(socket_info, socket_info->send_buffer,
+				  strlen(socket_info->send_buffer), false, false);
+
+		if (bytes < 0) {
+			/* Wait for socket to allow sending again */
+			socket_info->send_poll = true;
+			return;
+		}
+		socket_info->send_bytes_sent += bytes;
+		socket_info->send_bytes_left -= bytes;
+
+		/* Print throughput stats every 10 seconds */
+		time_intermediate = k_uptime_get();
+		ul_time_intermediate_ms = time_intermediate - socket_info->start_time_ms;
+
+		if ((ul_time_intermediate_ms / (double)1000) > socket_info->send_print_interval) {
+			throughput = calculate_throughput(
+				socket_info->send_bytes_sent,
+				ul_time_intermediate_ms);
+
+			shell_print(
+				shell_global,
+				"%7u bytes, %6.2fs, %6.0f bit/s",
+				socket_info->send_bytes_sent,
+				(float)ul_time_intermediate_ms / 1000,
+				throughput);
+
+			socket_info->send_print_interval += 10;
+		}
+	}
+	socket_info->send_poll = false;
+	ul_time_ms = k_uptime_delta(&socket_info->start_time_ms);
+	sock_send_buffer_free(socket_info);
+	print_throughput_summary(socket_info->send_bytes_sent, ul_time_ms);
+}
+
+int sock_send_data(
+	int socket_id,
+	char *data,
+	int random_data_length,
+	int interval,
+	bool blocking,
+	int buffer_size,
+	bool data_format_hex)
+{
+	struct sock_info_t *socket_info = get_socket_info_by_id(socket_id);
+	/* Process data to be sent based on input parameters */
+	int data_out_length = strlen(data);
+	uint8_t *data_out = data;
+	uint8_t data_out_hex[SOCK_MAX_SEND_DATA_LEN / 2 + 1] = { 0 };
```

Comment:
How big could this be? If either the thread stack or this value are configurable, we should leave a note to inform the users otherwise they could overflow the stack.

Reply:
SOCK_MAX_SEND_DATA_LEN=200 at the moment. As discussed in the meeting this is fine.